File: frontend/src/lib/api/isis.ts
/**
 * IS-IS Protocol API Service
 * Handles all IS-IS (Intermediate System to Intermediate System) related API operations
 */

import { apiClient } from "./client";

// ============================================================================
// Types
// ============================================================================

export interface ISISInterface {
  name: string;
  passive: boolean;
  circuit_type?: string | null;  // level-1, level-2, level-1-2
  metric?: number | null;
  priority?: number | null;
  hello_interval?: number | null;
  hello_multiplier?: number | null;
  network?: string | null;  // point-to-point
  bfd: boolean;
}

export interface ISISRedistribution {
  level: string;
  protocol: string;
  route_map?: string | null;
  metric?: number | null;
}

export interface ISISSPFDelay {
  init_delay?: number | null;
  short_delay?: number | null;
  long_delay?: number | null;
  holddown?: number | null;
  time_to_learn?: number | null;
}

export interface ISISConfig {
  configured: boolean;
  net: string[];
  is_type?: string | null;
  interfaces: ISISInterface[];
  redistributions: ISISRedistribution[];
  dynamic_hostname: boolean;
  metric_style?: string | null;
  lsp_mtu?: number | null;
  lsp_gen_interval?: number | null;
  lsp_refresh_interval?: number | null;
  max_lsp_lifetime?: number | null;
  set_attached_bit: boolean;
  set_overload_bit: boolean;
  purge_originator: boolean;
  spf_delay?: ISISSPFDelay | null;
}

export interface ISISCapabilities {
  is_types: { value: string; label: string; description: string }[];
  circuit_types: { value: string; label: string }[];
  metric_styles: { value: string; label: string; description: string }[];
  redistribute_protocols: { value: string; label: string }[];
  levels: { value: string; label: string }[];
}

export interface ApiResponse {
  success: boolean;
  message: string;
  error?: string;
}

// ============================================================================
// Service
// ============================================================================

class ISISService {
  /**
   * Get IS-IS configuration
   */
  async getConfig(): Promise<ISISConfig> {
    return apiClient.get<ISISConfig>("/vyos/protocols/isis/config");
  }

  /**
   * Get IS-IS capabilities
   */
  getCapabilities(): ISISCapabilities {
    return {
      is_types: [
        { value: "level-1", label: "Level 1", description: "Intra-area routing" },
        { value: "level-2", label: "Level 2", description: "Inter-area routing" },
        { value: "level-1-2", label: "Level 1-2", description: "Both intra and inter-area" },
      ],
      circuit_types: [
        { value: "level-1", label: "Level 1" },
        { value: "level-2", label: "Level 2" },
        { value: "level-1-2", label: "Level 1-2" },
      ],
      metric_styles: [
        { value: "narrow", label: "Narrow", description: "Original 6-bit metric" },
        { value: "wide", label: "Wide", description: "Extended 24-bit metric" },
        { value: "transition", label: "Transition", description: "Support both narrow and wide" },
      ],
      redistribute_protocols: [
        { value: "bgp", label: "BGP" },
        { value: "connected", label: "Connected" },
        { value: "kernel", label: "Kernel" },
        { value: "ospf", label: "OSPF" },
        { value: "static", label: "Static" },
      ],
      levels: [
        { value: "level-1", label: "Level 1" },
        { value: "level-2", label: "Level 2" },
      ],
    };
  }

  /**
   * Enable IS-IS
   */
  async enable(options: {
    net: string;
    isType?: string;
    dynamicHostname?: boolean;
    metricStyle?: string;
  }): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/isis/enable", {
      net: options.net,
      is_type: options.isType,
      dynamic_hostname: options.dynamicHostname ?? false,
      metric_style: options.metricStyle,
    });
  }

  /**
   * Disable IS-IS
   */
  async disable(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/isis/disable");
  }

  // =========================================================================
  // NET operations
  // =========================================================================

  async addNET(net: string): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/isis/net", { net });
  }

  async removeNET(net: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/isis/net/${encodeURIComponent(net)}`);
  }

  // =========================================================================
  // Interface operations
  // =========================================================================

  async configureInterface(
    interfaceName: string,
    options?: {
      passive?: boolean;
      circuitType?: string;
      metric?: number;
      priority?: number;
      helloInterval?: number;
      helloMultiplier?: number;
      network?: string;
      bfd?: boolean;
    }
  ): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/isis/interface", {
      interface: interfaceName,
      passive: options?.passive ?? false,
      circuit_type: options?.circuitType,
      metric: options?.metric,
      priority: options?.priority,
      hello_interval: options?.helloInterval,
      hello_multiplier: options?.helloMultiplier,
      network: options?.network,
      bfd: options?.bfd ?? false,
    });
  }

  async removeInterface(interfaceName: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/isis/interface/${encodeURIComponent(interfaceName)}`);
  }

  // =========================================================================
  // Redistribution operations
  // =========================================================================

  async addRedistribution(
    level: string,
    protocol: string,
    options?: { routeMap?: string; metric?: number }
  ): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/isis/redistribute", {
      level,
      protocol,
      route_map: options?.routeMap,
      metric: options?.metric,
    });
  }

  async removeRedistribution(level: string, protocol: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/isis/redistribute/${encodeURIComponent(level)}/${encodeURIComponent(protocol)}`);
  }

  // =========================================================================
  // IS Type operations
  // =========================================================================

  async setISType(isType: string): Promise<ApiResponse> {
    return apiClient.put<ApiResponse>(`/vyos/protocols/isis/is-type/${encodeURIComponent(isType)}`, {});
  }

  // =========================================================================
  // Feature toggles
  // =========================================================================

  async enableDynamicHostname(): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/isis/dynamic-hostname", {});
  }

  async disableDynamicHostname(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/isis/dynamic-hostname");
  }

  async enableOverloadBit(): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/isis/set-overload-bit", {});
  }

  async disableOverloadBit(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/isis/set-overload-bit");
  }

  // =========================================================================
  // LSP operations
  // =========================================================================

  async setLSPParameters(params: {
    lspMtu?: number;
    lspGenInterval?: number;
    lspRefreshInterval?: number;
    maxLspLifetime?: number;
  }): Promise<ApiResponse> {
    return apiClient.put<ApiResponse>("/vyos/protocols/isis/lsp-parameters", {
      lsp_mtu: params.lspMtu,
      lsp_gen_interval: params.lspGenInterval,
      lsp_refresh_interval: params.lspRefreshInterval,
      max_lsp_lifetime: params.maxLspLifetime,
    });
  }

  // =========================================================================
  // SPF Delay operations
  // =========================================================================

  async setSPFDelay(params: {
    initDelay?: number;
    shortDelay?: number;
    longDelay?: number;
    holddown?: number;
    timeToLearn?: number;
  }): Promise<ApiResponse> {
    return apiClient.put<ApiResponse>("/vyos/protocols/isis/spf-delay", {
      init_delay: params.initDelay,
      short_delay: params.shortDelay,
      long_delay: params.longDelay,
      holddown: params.holddown,
      time_to_learn: params.timeToLearn,
    });
  }

  async resetSPFDelay(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/isis/spf-delay");
  }
}

export const isisService = new ISISService();

File: backend/vyos_mappers/protocols/isis.py
"""
IS-IS Protocol Command Mapper

Handles IS-IS (Intermediate System to Intermediate System) configuration commands.
IS-IS is a link-state routing protocol for large networks.
"""

from typing import List, Dict, Any
from ..base import BaseFeatureMapper


class ISISMapper(BaseFeatureMapper):
    """IS-IS protocol mapper with all IS-IS configuration operations"""

    def __init__(self, version: str):
        """Initialize with VyOS version."""
        super().__init__(version)

    # ========================================================================
    # Basic IS-IS Configuration Paths
    # ========================================================================

    def get_isis(self) -> List[str]:
        """Get command path for creating IS-IS."""
        return ["protocols", "isis"]

    def get_net(self, net: str) -> List[str]:
        """Get command path for setting NET (Network Entity Title)."""
        return ["protocols", "isis", "net", net]

    def get_is_type(self, is_type: str) -> List[str]:
        """Get command path for IS type (level-1, level-2, level-1-2)."""
        return ["protocols", "isis", "is-type", is_type]

    def get_is_type_path(self) -> List[str]:
        """Get command path for IS type (for deletion)."""
        return ["protocols", "isis", "is-type"]

    def get_dynamic_hostname(self) -> List[str]:
        """Get command path for dynamic hostname."""
        return ["protocols", "isis", "dynamic-hostname"]

    def get_level(self, level: str) -> List[str]:
        """Get command path for level configuration."""
        return ["protocols", "isis", "level", level]

    def get_lsp_gen_interval(self, interval: str) -> List[str]:
        """Get command path for LSP generation interval."""
        return ["protocols", "isis", "lsp-gen-interval", interval]

    def get_lsp_mtu(self, mtu: str) -> List[str]:
        """Get command path for LSP MTU."""
        return ["protocols", "isis", "lsp-mtu", mtu]

    def get_lsp_refresh_interval(self, interval: str) -> List[str]:
        """Get command path for LSP refresh interval."""
        return ["protocols", "isis", "lsp-refresh-interval", interval]

    def get_max_lsp_lifetime(self, lifetime: str) -> List[str]:
        """Get command path for max LSP lifetime."""
        return ["protocols", "isis", "max-lsp-lifetime", lifetime]

    def get_metric_style(self, style: str) -> List[str]:
        """Get command path for metric style (narrow, wide, transition)."""
        return ["protocols", "isis", "metric-style", style]

    def get_set_attached_bit(self) -> List[str]:
        """Get command path for set-attached-bit."""
        return ["protocols", "isis", "set-attached-bit"]

    def get_set_overload_bit(self) -> List[str]:
        """Get command path for set-overload-bit."""
        return ["protocols", "isis", "set-overload-bit"]

    def get_purge_originator(self) -> List[str]:
        """Get command path for purge-originator."""
        return ["protocols", "isis", "purge-originator"]

    # ========================================================================
    # Interface Configuration Paths
    # ========================================================================

    def get_interface(self, interface: str) -> List[str]:
        """Get command path for interface configuration."""
        return ["protocols", "isis", "interface", interface]

    def get_interface_passive(self, interface: str) -> List[str]:
        """Get command path for passive interface."""
        return ["protocols", "isis", "interface", interface, "passive"]

    def get_interface_circuit_type(self, interface: str, circuit_type: str) -> List[str]:
        """Get command path for interface circuit type."""
        return ["protocols", "isis", "interface", interface, "circuit-type", circuit_type]

    def get_interface_hello_interval(self, interface: str, interval: str) -> List[str]:
        """Get command path for hello interval."""
        return ["protocols", "isis", "interface", interface, "hello-interval", interval]

    def get_interface_hello_multiplier(self, interface: str, multiplier: str) -> List[str]:
        """Get command path for hello multiplier."""
        return ["protocols", "isis", "interface", interface, "hello-multiplier", multiplier]

    def get_interface_metric(self, interface: str, metric: str) -> List[str]:
        """Get command path for interface metric."""
        return ["protocols", "isis", "interface", interface, "metric", metric]

    def get_interface_priority(self, interface: str, priority: str) -> List[str]:
        """Get command path for interface priority."""
        return ["protocols", "isis", "interface", interface, "priority", priority]

    def get_interface_network_point_to_point(self, interface: str) -> List[str]:
        """Get command path for point-to-point network type."""
        return ["protocols", "isis", "interface", interface, "network", "point-to-point"]

    def get_interface_bfd(self, interface: str) -> List[str]:
        """Get command path for BFD on interface."""
        return ["protocols", "isis", "interface", interface, "bfd"]

    # ========================================================================
    # Redistribution Configuration Paths
    # ========================================================================

    def get_redistribute(self, level: str, protocol: str) -> List[str]:
        """Get command path for redistributing a protocol."""
        return ["protocols", "isis", "redistribute", level, protocol]

    def get_redistribute_route_map(self, level: str, protocol: str, route_map: str) -> List[str]:
        """Get command path for redistribution route-map."""
        return ["protocols", "isis", "redistribute", level, protocol, "route-map", route_map]

    def get_redistribute_metric(self, level: str, protocol: str, metric: str) -> List[str]:
        """Get command path for redistribution metric."""
        return ["protocols", "isis", "redistribute", level, protocol, "metric", metric]

    # ========================================================================
    # Default Information Configuration
    # ========================================================================

    def get_default_information_originate(self, level: str) -> List[str]:
        """Get command path for originating default information."""
        return ["protocols", "isis", "default-information", "originate", level]

    def get_default_information_originate_always(self, level: str) -> List[str]:
        """Get command path for always originating default."""
        return ["protocols", "isis", "default-information", "originate", level, "always"]

    def get_default_information_originate_metric(self, level: str, metric: str) -> List[str]:
        """Get command path for default information metric."""
        return ["protocols", "isis", "default-information", "originate", level, "metric", metric]

    def get_default_information_originate_route_map(self, level: str, route_map: str) -> List[str]:
        """Get command path for default information route-map."""
        return ["protocols", "isis", "default-information", "originate", level, "route-map", route_map]

    # ========================================================================
    # SPF Configuration Paths
    # ========================================================================

    def get_spf_delay_ietf_init_delay(self, delay: str) -> List[str]:
        """Get command path for SPF initial delay."""
        return ["protocols", "isis", "spf-delay-ietf", "init-delay", delay]

    def get_spf_delay_ietf_short_delay(self, delay: str) -> List[str]:
        """Get command path for SPF short delay."""
        return ["protocols", "isis", "spf-delay-ietf", "short-delay", delay]

    def get_spf_delay_ietf_long_delay(self, delay: str) -> List[str]:
        """Get command path for SPF long delay."""
        return ["protocols", "isis", "spf-delay-ietf", "long-delay", delay]

    def get_spf_delay_ietf_holddown(self, delay: str) -> List[str]:
        """Get command path for SPF holddown."""
        return ["protocols", "isis", "spf-delay-ietf", "holddown", delay]

    def get_spf_delay_ietf_time_to_learn(self, time: str) -> List[str]:
        """Get command path for SPF time to learn."""
        return ["protocols", "isis", "spf-delay-ietf", "time-to-learn", time]

    # ========================================================================
    # Config Parsing Methods (for READ operations)
    # ========================================================================

    def parse_isis_config(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Parse full IS-IS configuration from VyOS.

        Args:
            config: Raw IS-IS config dictionary from VyOS (under protocols > isis)

        Returns:
            Parsed IS-IS configuration as dictionary
        """
        if not config:
            return {
                "configured": False,
                "net": [],
                "is_type": None,
                "interfaces": [],
                "redistributions": [],
                "dynamic_hostname": False,
                "metric_style": None,
                "lsp_mtu": None,
                "lsp_gen_interval": None,
                "lsp_refresh_interval": None,
                "max_lsp_lifetime": None,
                "set_attached_bit": False,
                "set_overload_bit": False,
                "purge_originator": False,
                "spf_delay": None,
            }

        # Parse NETs
        nets = []
        net_cfg = config.get("net", [])
        if isinstance(net_cfg, list):
            nets = net_cfg
        elif isinstance(net_cfg, dict):
            nets = list(net_cfg.keys())
        elif isinstance(net_cfg, str):
            nets = [net_cfg]

        # Parse interfaces
        interfaces = self._parse_interfaces(config.get("interface", {}))

        # Parse redistributions
        redistributions = self._parse_redistributions(config.get("redistribute", {}))

        # Parse SPF delay
        spf_delay = self._parse_spf_delay(config.get("spf-delay-ietf", {}))

        return {
            "configured": True,
            "net": nets,
            "is_type": config.get("is-type"),
            "interfaces": interfaces,
            "redistributions": redistributions,
            "dynamic_hostname": "dynamic-hostname" in config,
            "metric_style": config.get("metric-style"),
            "lsp_mtu": config.get("lsp-mtu"),
            "lsp_gen_interval": config.get("lsp-gen-interval"),
            "lsp_refresh_interval": config.get("lsp-refresh-interval"),
            "max_lsp_lifetime": config.get("max-lsp-lifetime"),
            "set_attached_bit": "set-attached-bit" in config,
            "set_overload_bit": "set-overload-bit" in config,
            "purge_originator": "purge-originator" in config,
            "spf_delay": spf_delay,
        }

    def _parse_interfaces(self, interfaces_config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse interface configurations."""
        interfaces = []
        for iface_name, iface_cfg in interfaces_config.items():
            if not isinstance(iface_cfg, dict):
                interfaces.append({
                    "name": iface_name,
                    "passive": False,
                    "circuit_type": None,
                    "metric": None,
                    "priority": None,
                    "hello_interval": None,
                    "hello_multiplier": None,
                    "network": None,
                    "bfd": False,
                })
                continue

            # Determine network type
            network_type = None
            network_cfg = iface_cfg.get("network", {})
            if isinstance(network_cfg, dict):
                if "point-to-point" in network_cfg:
                    network_type = "point-to-point"

            interfaces.append({
                "name": iface_name,
                "passive": "passive" in iface_cfg,
                "circuit_type": iface_cfg.get("circuit-type"),
                "metric": iface_cfg.get("metric"),
                "priority": iface_cfg.get("priority"),
                "hello_interval": iface_cfg.get("hello-interval"),
                "hello_multiplier": iface_cfg.get("hello-multiplier"),
                "network": network_type,
                "bfd": "bfd" in iface_cfg,
            })

        return interfaces

    def _parse_redistributions(self, redist_config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse redistribution configurations."""
        redistributions = []
        for level, level_cfg in redist_config.items():
            if not isinstance(level_cfg, dict):
                continue
            for protocol, protocol_cfg in level_cfg.items():
                if isinstance(protocol_cfg, dict):
                    redistributions.append({
                        "level": level,
                        "protocol": protocol,
                        "route_map": protocol_cfg.get("route-map"),
                        "metric": protocol_cfg.get("metric"),
                    })
                else:
                    redistributions.append({
                        "level": level,
                        "protocol": protocol,
                        "route_map": None,
                        "metric": None,
                    })

        return redistributions

    def _parse_spf_delay(self, spf_config: Dict[str, Any]) -> Dict[str, Any] | None:
        """Parse SPF delay configurations."""
        if not spf_config:
            return None

        return {
            "init_delay": spf_config.get("init-delay"),
            "short_delay": spf_config.get("short-delay"),
            "long_delay": spf_config.get("long-delay"),
            "holddown": spf_config.get("holddown"),
            "time_to_learn": spf_config.get("time-to-learn"),
        }
